src/modules/callouts/pages/CreateCallout.vue

```diff
@@ -144,77 +176,25 @@ const steps: Steps = reactive({
     },
   },
 });
-
-// TODO: should just be a computed
-watch(
-  () => steps.titleAndImage.data.title,
-  (title) => {
-    steps.url.data.autoSlug = slugify(title, { lower: true });
-  }
-);
-
-const stepsInOrder = computed(() => [
-  steps.content,
-  steps.titleAndImage,
-  steps.visibility,
-  steps.endMessage,
-  steps.url,
-  //steps.mailchimp,
-  steps.dates,
-]);
-
-const selectedStepIndex = ref(0);
-const selectedStep = computed(
-  () => stepsInOrder.value[selectedStepIndex.value]
-);
-const isLastStep = computed(
-  () => selectedStepIndex.value === stepsInOrder.value.length - 1
-);
-const isAllValid = computed(() =>
-  stepsInOrder.value.every((step) => step.validated)
-);
-
 async function submitForm() {
-  const callout: CreateCalloutData = {
-    slug: steps.url.data.useCustomSlug
-      ? steps.url.data.slug
-      : steps.url.data.autoSlug,
-    title: steps.titleAndImage.data.title,
-    excerpt: steps.titleAndImage.data.description,
-    image: steps.titleAndImage.data.coverImageURL,
-    intro: steps.content.data.introText,
-    formSchema: steps.content.data.formSchema,
-    starts: steps.dates.data.startNow
-      ? new Date()
-      : parseISO(steps.dates.data.startDate),
-    ...(steps.dates.data.hasEndDate && {
-      expires: parseISO(steps.dates.data.endDate),
-    }),
-    allowUpdate: steps.visibility.data.usersCanEditAnswers,
-    allowMultiple: false,
-    hidden: !steps.visibility.data.showOnUserDashboards,
-    access:
-      steps.visibility.data.whoCanTakePart === 'members'
-        ? 'member'
-        : steps.visibility.data.allowAnonymousResponses
-        ? 'anonymous'
-        : 'guest',
-    ...(steps.endMessage.data.whenFinished === 'message'
-      ? {
-          thanksText: steps.endMessage.data.thankYouText,
-          thanksTitle: steps.endMessage.data.thankYouTitle,
-        }
-      : {
-          thanksText: '',
-          thanksTitle: '',
-          thanksRedirect: steps.endMessage.data.thankYouRedirect,
-        }),
-  };
+  // @ts-expect-error
+  const callout: CreateCalloutData = makeCalloutData(steps);
 
-  const newCallout = await createCallout(callout);
+  if (mode === 'edit') {
+    const newCallout = await editCallout(callout);
+  } else if (mode === 'new') {
+    const newCallout = await createCallout(callout);
+  }
   router.push({
-    path: '/admin/callouts/edit/' + newCallout.slug,
+    path: '/admin/callouts/edit/' + callout.slug,
     query: { created: null },
   });
 }
+onBeforeMount(async () => {
+  const calloutId = route.params.id as string;
```

Comment:
This method should use either the `async` pattern or `Promise.then().catch()` (I'd prefer `async`/`await`)

I don't think it should default to an empty form if the callout doesn't exist so I wouldn't bother catching the error. Ideally it would go to a 404 page but we don't actually have one of those at the moment :cold_sweat: 

The code could be simplified to something like:
```ts
steps.value = mode === 'edit'
  ? makeStepsData(await fetchCallout(calloutId))
  : makeStepsData({})
```

Reply:
I'm surprised that `{}` passes the type checker, but it would probably be more explicit if `makeStepsData` could accept no parameter (i.e. make the parameter optional)